registry/handlers/referrers.go

```diff
@@ -0,0 +1,196 @@
+package handlers
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"path"
+
+	"github.com/distribution/distribution/v3"
+	dcontext "github.com/distribution/distribution/v3/context"
+	"github.com/distribution/distribution/v3/manifest/ociartifact"
+	"github.com/distribution/distribution/v3/registry/api/errcode"
+	v2 "github.com/distribution/distribution/v3/registry/api/v2"
+	"github.com/distribution/distribution/v3/registry/storage"
+	"github.com/distribution/distribution/v3/registry/storage/driver"
+	"github.com/gorilla/handlers"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/specs-go"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// referrersDispatcher takes the request context and builds the
+// appropriate handler for handling referrers requests.
+func referrersDispatcher(ctx *Context, r *http.Request) http.Handler {
+	dgst, err := getDigest(ctx)
+	if err != nil {
+		if err == errDigestNotAvailable {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				ctx.Errors = append(ctx.Errors, v2.ErrorCodeDigestInvalid.WithDetail(err))
+			})
+		}
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx.Errors = append(ctx.Errors, v2.ErrorCodeDigestInvalid.WithDetail(err))
+		})
+	}
+
+	referrersHandler := &referrersHandler{
+		Context: ctx,
+		Digest:  dgst,
+	}
+	return handlers.MethodHandler{
+		"GET": http.HandlerFunc(referrersHandler.GetReferrers),
+	}
+}
+
+// referrersHandler handles http operations on referrers.
+type referrersHandler struct {
+	*Context
+	Digest digest.Digest
+}
+
+// GetReferrers fetches the list of referrers as an image index from the storage.
+func (h *referrersHandler) GetReferrers(w http.ResponseWriter, r *http.Request) {
+	dcontext.GetLogger(h).Debug("GetReferrers")
+
+	if h.Digest == "" {
+		h.Errors = append(h.Errors, v2.ErrorCodeManifestUnknown.WithDetail("digest not specified"))
+		return
+	}
+
+	annotations := map[string]string{}
+	query := r.URL.Query()
+	artifactTypeFilter := query.Get("artifactType")
+	if artifactTypeFilter != "" {
+		annotations[v1.AnnotationReferrersFiltersApplied] = "artifactType"
+	}
```

Comment:
note that `nil` map with `omitempty` will be omitted. However, empty map will be encoded to `{}`.

Here, if `artifactType` is not there, we should use `nil` map.

Consider this code pattern
```suggestion
	var annotations map[string]string{}
	if artifactTypeFilter := r.URL.Query().Get("artifactType"); artifactTypeFilter != "" {
		annotations = map[string]string{
			v1.AnnotationReferrersFiltersApplied: "artifactType",
		}
	}
```

Reply:
changed accordingly.